Module2

sonasol2 opened the pull request:
-просмотрел материалы и постарался привести код в соответствие с рекомендациями
-задокументировал интерфейсы и модели
-try catch сделал в program
-постарался полнее сымитировать бд, сделал статической коллекцию и добавил методы с локами для работы с бд
-сделал базовую реализацию методов в Repository(пока так, потом когда EF подключим я сделаю ее полноценнее, сейчас сложновато из-за того что кастомная бд)
-сделал другой middleware для глобальной обработки исключений, увидел что такой подход рекомендуют
-раздлелил доменные модели и модели сущности для DAL слоя
-разделил сущности на Region, Advertising и AdvertisingPlatform
-зарегестрировал все в DI как singletone
-добавил дату публикации в Advertising и расширил поиск с помощью делегата

sonasol2 committed:
update

add
-domain models
-base abstract class repository
-logger in global exception handler

update
-name for entity in DAL
-name folder for abstractions
-name repository

sonasol2 committed:
update

add
-domain model abstraction
-factories abstractions
-concrete factories
-concrete repositories

update
-name domain model
-cacheContext collection is static now
-repository base implementation

sonasol2 committed:
update

add:
-new global exception handler
-error response model
-documentation for iRepository

update:
-services to singleton

remove:
-return ExecutionResult in advertisingRegionService

sonasol2 committed:
update

add
-request/response models for controller
-automation validate

update
-now model request in controller instead of simple value
-validation based on request model
-validate logic

remove
-validate service

sonasol2 committed:
update

add
-dto
-delegate for search with parameters
-class mapper in API
-factory for generation search parameters

update
-thread safe in cacheContext
-now search in service with many parameters
-add createAll method in factory

delete
-ExecutionResult
-iValidatorService

sonasol2 committed:
update

add
-documentation for public interface and models
-try catch in program

fix
-little rename

delete
-IPlatform

sonasol2 committed:
fix search

Vadya0994 commented on AdvertisingRegionService.API/Controllers/AdvertisingPlatformsController.cs:
```diff
@@ -9,46 +10,38 @@ namespace AdvertisingRegionService.API.Controllers;
         
-        return Ok(LocalizationConstants.SuccessMessage);
+        if (!result)
+            return BadRequest();
```
Заведи свой тип исключений, унаследованный от System.Exception и бросай там в методе UploadFile, потом можешь в обработчике исключений отлавливать и заворачивать в ошибку, так все контроллеры будут чистыми и не будут содержать лишнюю логику

sonasol2 replied:
Хорошо! Сделал кастомное исключение и генерирую его в методе сервисе

Vadya0994 commented on AdvertisingRegionService.API/Controllers/AdvertisingPlatformsController.cs:
```diff
@@ -9,46 +10,38 @@ namespace AdvertisingRegionService.API.Controllers;
-        if(!validateResult.Result.IsValid) return BadRequest(validateResult.Result.Errors);
+        var searchPredicates = PredicateSearchFactory.CreateSearchPredicates(searchRequest);
+        var result = _advertisingRegionService.SearchPlatform(searchPredicates);
+        if (result == null || !result.Any())
```
И тут точно также

sonasol2 replied:
А я подумал, можно ли, вместо этого не исключение отлавливать в сервисе, а возвращать просто пустую коллекцию??  По типу того? Я подумал о том, что .Any() тоже вроде как по коллекции проходит?
`
 return filteredPlatforms.Count == 0 
            ? new List<AdvertisingPlatform>()
            : filteredPlatforms; 
`

Vadya0994 commented on AdvertisingRegionService.API/Middlewares/GlobalExceptionHandler.cs:
```diff
@@ -0,0 +1,63 @@
+        switch (exception)
+        {
+            case ArgumentNullException:
+            case NullReferenceException:
```
Поидее NullReferenceException тянет на InternalServerError, его лучше возвращать для всех необработанных исключений

sonasol2 replied:
Убрал

Vadya0994 commented on AdvertisingRegionService.API/Middlewares/GlobalExceptionHandler.cs:
```diff
@@ -0,0 +1,63 @@
+                response.Message = exception.Message;
+                break;
+            case KeyNotFoundException:
+                response.StatusCode = HttpStatusCode.NotFound;
```
А этот тип исключений, когда ресурс не найден (когда URL не правильно написал), лучше бизнес логикой такое разрулить и венуть пустую коллекцию или бросить свое кастомное бизнес исключение в участке когда, где ожидаешь такую проблему

sonasol2 replied:
Убрал

Vadya0994 commented on AdvertisingRegionService.API/Middlewares/GlobalExceptionHandler.cs:
```diff
@@ -0,0 +1,63 @@
+                response.StatusCode = HttpStatusCode.BadRequest;
+                response.Message = exception.Message;
+                break;
+            case ArgumentException:
```
да и это тоже не нужно обрабатывать)

sonasol2 replied:
Убрал

Vadya0994 commented on AdvertisingRegionService.API/Middlewares/GlobalExceptionHandler.cs:
```diff
@@ -0,0 +1,63 @@
+            default:
+                response.StatusCode = HttpStatusCode.InternalServerError;
+                response.Message = exception.Message;
+                break;
```
Кстать в режиме разработки неплохо бы возвращать stack trace исключений, чтобы понимать где искать проблему, когда не ожидаешь)

sonasol2 replied:
Добавил!

Vadya0994 commented on AdvertisingRegionService.API/Validators/SearchValidator.cs:
```diff
@@ -1,20 +1,20 @@
+            .NotNull().WithMessage(LocalizationConstants.NullModelMessage)
+            .NotEmpty().WithMessage(LocalizationConstants.EmptySearchStringMessage)
+            .Length(ValidationConstants.MinSearchLength, ValidationConstants.MaxSearchLength)
+            .WithMessage($"Lenght must be at {ValidationConstants.MinSearchLength} to {ValidationConstants.MaxSearchLength} symbols");
```
А почему все сообщения в константах, а это нет?)

sonasol2 replied:
Показалось что так логично объединить две константы:)) Переделал!

Vadya0994 commented on AdvertisingRegionService.DAL/Abstractions/IRepository.cs:
```diff
@@ -0,0 +1,40 @@
+    /// Get all entities of type <typeparam name="T"/>
+    /// </summary>
+    /// <returns>Query <see cref="IQueryable{T}"/> to select all entities</returns>
+    IQueryable<T> GetAll();
```
Возвращать запрос из репозитория плохая идея. Репозиторий должен формировать запросы и материализовывать данные в оперативную память. Тут понятно, что они у тебя и так в памяти, но прятать под query это также плохая идея

sonasol2 replied:
Понял, поменял на IReadOnlyCollection

Vadya0994 commented on AdvertisingRegionService.DAL/Repositories/AdvertisingPlatformRepository.cs:
```diff
@@ -0,0 +1,40 @@
+        _cacheContext.AdvertisingPlatform.Clear();
+    }
+
+    public HashSet<string> GetAdvertisingPlatformByRegion(string region)
```
Где-то в названии методов можно не перебарщивать, например здесь GetAdvertisingPlatformByRegion можно назвать просто GetByRegion. Так как и так понятно, что ты получаешь AdvertisingPlatform, у тебя класс называется AdvertisingPlatformRepository

sonasol2 replied:
Логично, переименовал!

Vadya0994 commented on AdvertisingRegionService.Domain/Abstractions/IAdvertisingRegionFileParser.cs:
```diff
@@ -0,0 +1,17 @@
+    /// </summary>
+    /// <param name="fileContents">The content of the text file as a string.</param>
+    /// <returns>A list of advertising platform</returns>
+    List<AdvertisingPlatformEntity> Parse(string fileContents);
```
Старайся по возможности больше оперировать интерфейсами + если не планируется по задумке изменять снаружи коллекцию, то используй readonly, я бы вернул тут какой-нибудь IReadonlyCollection, это все добавляет более выразительную читаемость, защиту данных снаружи, что нигде не добавишь ничего в эту коллекцию, ну и работа с интерфейсами - слабосвязанность кода

sonasol2 replied:
Понял! С коллекциями постоянно путаюсь что где лучше использовать, буду внимательнее.

Vadya0994 commented on AdvertisingRegionService.Domain/Abstractions/IAdvertisingRegionService.cs:
```diff
@@ -0,0 +1,25 @@
+    /// </summary>
+    /// <param name="searchPredicates">A list of parameters defines the search criteria. </param>
+    /// <returns>An IEnumerable of <see cref="IAdvertisingPlatform"/> objects matching the search criteria, or null if no platforms are found.</returns>
+    IEnumerable<IAdvertisingPlatform>? SearchPlatform(List<SearchPredicate> searchPredicates);
```
Если снаружи, нужны такие методы как Count или Any - то лучше не прятать под IEnumerable, так как будешь вводить в заблуждение других программистов, работавшими с этим методом, они то могут и не знать, что там спрятана коллекция, а не итератор.

Vadya0994 replied:
Поглядел, что у тебя там в этом методе ниже. 2 раза перечисляешь итерируемую коллекцию, что дает просадку по производительности. Думаю что ее лучше материлизовать, либо сделать так - чтобы перечислялась один раз снаружи

Vadya0994 commented on AdvertisingRegionService.Domain/DTO/AdvertisingDto.cs:
```diff
@@ -0,0 +1,7 @@
+namespace AdvertisingRegionService.Domain.DTO;
+
+public class AdvertisingDto
```
Зачем эти ДТО? Они нигде не используются и во вторых у тебя на уровне домена, работа с доменными моделями, ДТО обычно в моделях контроллера используются, которые конвертируются из доменных моделей и используются для передачи данных по сети

sonasol2 replied:
Подумал что буду их использовать как объекты для передачи данных между слоями, а по итогу подумал оставить на будущее:))) Убрал их

Vadya0994 commented on AdvertisingRegionService.Domain/DTO/AdvertisingPlatformDto.cs:
```diff
@@ -0,0 +1,8 @@
+namespace AdvertisingRegionService.Domain.DTO;
+
+public class AdvertisingPlatformDto
```
ну и вот эти 2 модели ниже

sonasol2 replied:
Убрал

Vadya0994 commented on AdvertisingRegionService.Domain/Factories/AdvertisingFactory.cs:
```diff
@@ -0,0 +1,32 @@
+        return new Advertising(entity);
+    }
+
+    public override IEnumerable<IAdvertising> CreateAll(IEnumerable<AdvertisingEntity> entities)
```
не прячь List под Ienumerable, верни IreadOnlyCollection

Vadya0994 replied:
И в других местах

Vadya0994 replied:
И я бы назвал метод CreateAny, выглядело бы более наглядно

sonasol2 replied:
Переименовал и исправил на IReadOnlyCollection

Vadya0994 commented on AdvertisingRegionService.Domain/Models/Advertising.cs:
```diff
@@ -0,0 +1,42 @@
+    public HashSet<string> AdvertisingName
+    {
+        get => _advertisingEntity.PlatformNames;
+        set =>  _advertisingEntity.PlatformNames = value;}
```
Я обычно на пробелы и переносы не обращаю внимания, но закрывающая скобка методы в конце строки очень нелепо выглядит)

sonasol2 replied:
Не доглядел, действительно выглядит отвратительно:)

Vadya0994 commented on AdvertisingRegionService.Domain/Parsers/AdvertisingRegionFileParser.cs:
```diff
@@ -37,16 +48,16 @@ private void AddPlatformToRegions(string platform, HashSet<string> regions, Dict
-                Region = region,
-                Platforms = GetPlatformsForRegionHierarchy(region, regionPlatforms)
+                Region = new RegionEntity() {Name = region},
+                Platform = new AdvertisingEntity() {PlatformNames = GetPlatformsForRegionHierarchy(region, regionPlatforms), PostedAt = DateTime.UtcNow} 
```
работу со структурой DateTime я бы сделал через свой кастомный IDateTimeHelper с заделом на будущее. Так в будущем можно будет просто тестировать функционал связанный со временем, плюс часовые пояса пользователей и возможно где-то что-то нужно будет выводить в локальном времени и у тебя будет единая точка управления всего этого в одном месте  и что-то быстро можно будет быстро поменять, а не искать по всему проекту.

sonasol2 replied:
Сделал! Пока что просто добавил два поля в него.

Vadya0994 commented on AdvertisingRegionService.Domain/Services/AdvertisingRegionService.cs:
```diff
@@ -25,17 +30,29 @@ public async Task<ExecutionResult<bool>> UploadFile(Stream file)
+    public IEnumerable<IAdvertisingPlatform>? SearchPlatform(List<SearchPredicate> searchPredicates)
     {
-        return ExecutionResult<HashSet<string>>.Success(_cacheRepository.GetAdvertisingPlatformByLocation(searchRequest));
+        var entityAdvertisingPlatforms = _advertisingPlatformRepository.GetAll().AsEnumerable();
```
тут конечно странные манипуляции, query преобразуешь в enumerable)

sonasol2 replied:
Теперь вижу что действительно странные:))

sonasol2 committed:
fix exception

update
- add stack trace to errorResponse
- delete unnecessary exception handler
- fix constant in validate
- remove unnecessary return in controller

sonasol2 committed:
fix collections

update
- returnable collections across the project

sonasol2 committed:
other fix

add
- dateHelper

update
- base repository implementation
- cache context method

delete
- dto
